### PR TITLE
Rework host/port and listener setup

### PR DESCRIPTION
**Pull request by alexrudd2**
The refactored code is less nested, so I hope it is easier to understand.

It also solves a `mypy` error.

### CODE REVIEW

**Comment by alexrudd2** [comment]
See https://github.com/pymodbus-dev/pymodbus/discussions/1865.

If the test that creates `host == socket://__pymodbus_nullmodem:7636` is not valid, this can be simplified further.

**Comment by janiversen** [comment]
> See #1865.
> 
> If the test that creates `host == socket://__pymodbus_nullmodem:7636` is not valid, this can be simplified further.

That is a valid specification, the definition is "socket://<host>:<port>" for serial port as sockets.

**Comment by alexrudd2** [comment]
> You are touching a very delicate part of the code.
> 
> Some of your does not seem correct at a first look,but might very well be correct.

It will remain a secret, how many test failures I had in my attempts 🤣

File: pymodbus/transport/transport.py [hunk 1]
```diff
@@ -176,39 +176,37 @@ def __init__(
         else:
             host = self.comm_params.host
             port = int(self.comm_params.port)
-        if self.comm_params.comm_type == CommType.SERIAL:
-            host, port = self.init_setup_serial(host, port)
-            if not host and not port:
-                return
-        if host == NULLMODEM_HOST:
+        if NULLMODEM_HOST in host:
```
Comment (janiversen) [inline pymodbus/transport/transport.py:179]: this allows "NULLMODEM_HOST:7109" for all commtype...and it is only defined for serial.

Not sure it is a big problem, but at least a concern.

File: pymodbus/transport/transport.py [hunk 2]
```diff
@@ -176,39 +176,37 @@ def __init__(
...
-            return NULLMODEM_HOST, int(host[9:].split(":")[1])
-        if self.is_server and host.startswith("socket"):
+        if (
+            self.comm_params.comm_type == CommType.SERIAL
+            and self.is_server
+            and host.startswith("socket")
+        ):
             # format is "socket://<host>:port"
             self.comm_params.comm_type = CommType.TCP
+            host = host.strip("socket://")
```
Comment (janiversen) [inline pymodbus/transport/transport.py:191]: the old code was more compact, and did the job, so why change it ?

**Review by janiversen — CHANGES_REQUESTED** [review]

File: pymodbus/transport/transport.py (hunk 2 above)
Comment (alexrudd2) [reply pymodbus/transport/transport.py:191]: clarity

File: pymodbus/transport/transport.py (hunk 1 above)
Comment (alexrudd2) [reply pymodbus/transport/transport.py:179]: This is functionally possible in the old/existing code as well:
https://github.com/pymodbus-dev/pymodbus/blob/6b1ee62b34c50f4e8f3f8c9a9b42e02ff2f8ffe4/pymodbus/transport/transport.py#L183-L185

It's certainly possible that was unintended, in which case the refactor exposed a bug :)

File: pymodbus/transport/transport.py (hunk 2 above)
Comment (janiversen) [reply pymodbus/transport/transport.py:191]: well that is not more clear to me...now I have to make sure nothing references host after it is changed.

Comment (alexrudd2) [reply pymodbus/transport/transport.py:191]: Would you prefer below?
```python
parts = host.strip("socket://").split(":")
```

File: pymodbus/transport/transport.py (hunk 1 above)
Comment (janiversen) [reply pymodbus/transport/transport.py:179]: no the old code checked for exact match, those lines would not accept "NULLMODEM:9657"

Comment (alexrudd2) [reply pymodbus/transport/transport.py:179]: d'oh, you're right

File: pymodbus/transport/transport.py (hunk 2 above)
Comment (janiversen) [reply pymodbus/transport/transport.py:191]: That is at least still a one liner, but I fail to understand what is wrong with the original code, that only indexed the string.

I am not in favor of making changes just for the sake of changing... months from now when bisecting you will wonder if there was a reason why it was changed and if the old code was faulty.

**Review by janiversen — APPROVED** [review]
LGTM.

**Comment by alexrudd2** [comment]
The reason I tried changing things here to was to avoid this inelegant return of `None, None`.

```python
if self.comm_params.comm_type == CommType.SERIAL:
        host, port = self.init_setup_serial(host, port)
        if not host and not port:
            return

def init_setup_serial(self, host: str, _port: int) -> tuple[str, int]:
...
return None, None
```

Perhaps you have a cleaner way.

**Comment by janiversen** [comment]
> The reason I tried changing things here to was to avoid this inelegant return of `None, None`.
> 
> ```python
> if self.comm_params.comm_type == CommType.SERIAL:
>         host, port = self.init_setup_serial(host, port)
>         if not host and not port:
>             return
> 
> def init_setup_serial(self, host: str, _port: int) -> tuple[str, int]:
> ...
> return None, None
> ```
> 
> Perhaps you have a cleaner way.

Your last change looks ok....but you still have the change with strip and split, which I overlooked when I first approved the PR.

File: pymodbus/transport/transport.py [hunk 3]
```diff
@@ -176,39 +176,36 @@ def __init__(
...
-            self.handle_new_connection,
-            host,
-            baudrate=self.comm_params.baudrate,
-            bytesize=self.comm_params.bytesize,
-            parity=self.comm_params.parity,
-            stopbits=self.comm_params.stopbits,
-            timeout=self.comm_params.timeout_connect,
-        )
-        return None, None
+            parts = host.strip("socket://").split(":")
```
[the remark this answers is not in the document]
Comment (alexrudd2) [reply pymodbus/transport/transport.py]:
```python
# Alex prefers being explicit
parts = host.strip("socket://").split(":")
host, port = parts[0], int(parts[1])

# jan prefers being compact
parts = host.split(":")
host, port = parts[1][2:], int(parts[2])
```

**Review by janiversen — APPROVED** [review]
LGTM, thanks.

**Pull request merged**